src/main/java/graphql/execution/ExecutionStrategy.java

```diff
@@ -120,25 +122,29 @@ protected ExecutionResult completeValue(ExecutionContext executionContext, Graph
         return executionContext.getQueryStrategy().execute(executionContext, resolvedType, result, subFields);
     }
 
-    private ExecutionResult completeValueForList(ExecutionContext executionContext, GraphQLList fieldType, List<Field> fields, Object result) {
+    private ExecutionResult completeValueForList(ExecutionContext executionContext, GraphQLList fieldType, List<Field> fields, Object result, Map<String, Object> argumentValues) {
         if (result.getClass().isArray()) {
             result = Arrays.asList((Object[]) result);
         }
 
         //noinspection unchecked
-        return completeValueForList(executionContext, fieldType, fields, (Iterable<Object>) result);
+        return completeValueForList(executionContext, fieldType, fields, (Iterable<Object>) result, argumentValues);
     }
 
-    protected GraphQLObjectType resolveType(GraphQLInterfaceType graphQLInterfaceType, Object value) {
-        GraphQLObjectType result = graphQLInterfaceType.getTypeResolver().getType(value);
+    protected GraphQLObjectType resolveType(Field field, GraphQLInterfaceType graphQLInterfaceType,
+                                            Object value, Map<String, Object> argumentValues, GraphQLSchema schema) {
+        TypeResolutionEnvironment env = new TypeResolutionEnvironment(value, argumentValues, field, graphQLInterfaceType, schema);
+        GraphQLObjectType result = graphQLInterfaceType.getTypeResolver().getType(env);
         if (result == null) {
             throw new GraphQLException("could not determine type");
         }
         return result;
     }
 
-    protected GraphQLObjectType resolveType(GraphQLUnionType graphQLUnionType, Object value) {
```

Comment:
The parameters pattern is simply taking N current parameters and moving them into an object with a builder so it takes only 1 parameters

eg:

    resolveField(Field field, GraphQLInterfaceType graphQLInterfaceType,
                                             Object value, Map<String, Object> argumentValues, GraphQLSchema schema)

becomes

   resolveField(ResolveFieldParameters parameters)

this is useful because you are able to "add" new parameters to the builder while preserving backwards compatibility.  This is more useful for "API /  SPI" of course

The following is for "constructors" from Effective Java

http://www.informit.com/articles/article.aspx?p=1216151&seqNum=2

But it would be for methods as well

Note this is not really your problem in terms of this PR.  But because you needed to add yet more parameters you exasperated the existing problem.

A parameters object would also be a breaking change, although we might be able to do it via overloading and deprecation